pkg/controller/endpointslice/reconciler.go

```diff
@@ -187,13 +187,15 @@ func (r *reconciler) finalize(
 		}
 		sliceToDelete := slicesToDelete[i]
 		slice := slicesToCreate[len(slicesToCreate)-1]
-		// Only update EndpointSlices that have the same AddressType as this
-		// field is considered immutable. Since Services also consider IPFamily
-		// immutable, the only case where this should matter will be the
-		// migration from IP to IPv4 and IPv6 AddressTypes, where there's a
+		// Only update EndpointSlices that are owned by this Service and have
+		// the same AddressType. We need to avoid updating EndpointSlices that
+		// are being garbage collected for an old Service with the same name.
+		// The AddressType field is immutable. Since Services also consider
+		// IPFamily immutable, the only case where this should matter will be
+		// the migration from IP to IPv4 and IPv6 AddressTypes, where there's a
 		// chance EndpointSlices with an IP AddressType would otherwise be
 		// updated to IPv4 or IPv6 without this check.
-		if sliceToDelete.AddressType == slice.AddressType {
+		if sliceToDelete.AddressType == slice.AddressType && ownedBy(sliceToDelete, service) {
```

Comment:
why not filtering all "not owned" slices in `reconcile`?

Reply:
Do you think that `syncService` should be a better place? so `reconcile()` and `finalize()` does not have to deal with "undesired" slices?

Reply:
That's a good idea, I'd thought about it as well but everything I tried ended up being messier/more complex than this change. Part of it is that syncService is intentionally small in scope and adding a check like this there ended up increasing the scope/complexity more than expected. Related to that, the majority of existing test coverage is focused on reconciler functions, so adding key functionality outside of that requires a larger investment in new testing.

Reply:
fair enough, I was thinking more in how endpoint slices consumers should implement this, and trying to come up with a "recommended" way of listing slices that belong to a service, based on this issue it seems that Listing based on label only is not enough.


```go
	esLabelSelector := labels.Set(map[string]string{
		discovery.LabelServiceName: service.Name,
		discovery.LabelManagedBy:   controllerName,
	}).AsSelectorPreValidated()
	endpointSlices, err := c.endpointSliceLister.EndpointSlices(service.Namespace).List(esLabelSelector)
```

Unrelated to this PR based on your comments but ... maybe we should offer some helper function that returns the slices that belong to a current service?

Reply:
That's a great point. I think listing slices based on label has to be sufficient simply because we don't and can't control all the ways EndpointSlices can be created. An owner ref is technically optional for producers here. I think this is more of a problem for producers to ensure they clean up old resources than for consumers. If/when this is approved, I'll have to make a very similar PR for the EndpointSliceMirroring controller.